### ISSUE TITLE
Events get mangled when non-utf8 value in Value

### ISSUE THREAD

**Issue by ethanfrey** [tendermint/tendermint#6403]
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):

v0.34.10

**ABCI app** (name for built-in, URL for self-written if it's publicly available):

Cosmos SDK v0.42.4 (and master)

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:

**What happened**:

A custom IBC app dared to actually return binary data in the acknowledgement, which was converted to a string `string(rawBytes)` for the cosmos sdk event, then back to `[]byte` for abci, and then into the tendermint rpc event system. Somewhere along the way it got mangled and a few bytes changed - which caused the protobuf data to not decode.

Relevant issue and fix (hex-encode binary before passing to event system) here: https://github.com/cosmos/ibc-go/pull/144

**What you expected to happen**:

If the abci.Event takes key and value as `[]byte` it should support them, no?

**Have you tried the latest version**: yes

**How to reproduce it** (as minimally and precisely as possible):

See https://github.com/cosmos/ibc-go/pull/144

**Suggestion**

Just make the `KVPair` used for returning Events over abci take `key string`, `value string`. There is no real reason for this and lots of code unsafely converts between strings and bytes. (I doubt any indexer handles raw binary well).

Once it used to be a nicely typed `string` -> `string | int` mapping

```proto
message KVPair {
  string key = 1;
  enum Type {
    STRING = 0;
    INT = 1;
  }
  Type value_type = 2;
  string value_string = 3;
  int64 value_int = 4;
}
```

But it was later converted to

```proto
message KVPair {
  bytes key = 1;
  bytes value = 2;
}
```

in https://github.com/tendermint/abci/pull/167. Although the arguments don't really convince me (as no processing code knows what to do with `[]byte` besides try to cast it):

> -> Protobuf3 strings are supposed to be only utf8 or 7-bit ansii but for something called "KVPair", we don't want to be restricted to unicode or ansii characters.
> -> Perhaps using []byte makes us more vigilant in validating the input before accepting it, than be lulled into complacency w/ a string.

This actually broke all searches in the event system, as it didn't know how to handle anything anymore. You can see a desire to revert this change here: https://github.com/tendermint/tendermint/issues/1369 (particularly comment https://github.com/tendermint/tendermint/issues/1369#issuecomment-377169038) and it seemed to have been decided to use `strings` again: https://github.com/tendermint/tendermint/issues/1369#issuecomment-383884334

Although the only fix was that the event system *expects all tags to be strings* https://github.com/tendermint/tendermint/pull/1498. It uses `map[string]string` to store the data that is abci-encoded as `[]byte`. WHY?

Please update the abci interface to use strings. This inspired refactoring is causing issues more than 3 years later.

**Comment by tessr** [comment]
Thanks for this detailed report. Just to make sure I understand: your recommendation is to change the ABCI interface for [`EventAttribute`](https://github.com/tendermint/tendermint/blob/master/proto/tendermint/abci/types.proto#L285) to use strings instead of bytes?

**Comment by alexanderbez** [comment]
> Thanks for this detailed report. Just to make sure I understand: your recommendation is to change the ABCI interface for [`EventAttribute`](https://github.com/tendermint/tendermint/blob/master/proto/tendermint/abci/types.proto#L285) to use strings instead of bytes?

Yeah I think so and I support the proposal!

**Comment by ethanfrey** [comment]
@tessr exactly. I request to change the ABCI interface for EventAttribute to use strings instead of bytes. That will at least avoid invalid data being being treated as strings.

If you want to go back towards typed events, I think that could be an interesting approach, but definitely a different issue/pr, and should be done as part of the larger event streaming/indexing changes

**Comment by alessio** [comment]
> > Thanks for this detailed report. Just to make sure I understand: your recommendation is to change the ABCI interface for [`EventAttribute`](https://github.com/tendermint/tendermint/blob/master/proto/tendermint/abci/types.proto#L285) to use strings instead of bytes?
> 
> Yeah I think so and I support the proposal!

FWIW I too support this proposal :+1:

**Comment by alexanderbez** [comment]
Posting here for visibility:

@ethanfrey should we break proto here by changing the types or should we introduce new types, `key_str` and `value_str`, and use those while keeping the existing ones deprecated?

**Comment by ethanfrey** [comment]
Personally, I would break proto (we are pre 1.0) and release this with tendermint 0.35, which will go along with a breaking Cosmos sdk change to match.

But I don't know if there are deeper implications. Like would this make it impossible to x/upgrade from a 0.34 to a 0.35 chain?

I feel it would be good to migrate this now before it gets forgotten again and not make it in 1.0

**Comment by alexanderbez** [comment]
I don't think this will require a migration. We already were casting to strings internally.

**Comment by ethanfrey** [comment]
Thanks Bez

**Comment by thanethomson** [comment]
Reopening since this the solution to this hasn't yet been ported across to `main`/v0.37+.

We've made a commitment now to no longer break the protos, and as such we may need a slightly different solution here to what was implemented in #6408.

**Comment by marbar3778** [comment]
> We've made a commitment now to no longer break the protos, and as such we may need a slightly different solution here to what was implemented in #6408.

Why the commitment to not breaking proto files? if the files are versioned this is more than fine.
